iina/FFmpegController.mm

```diff
@@ -0,0 +1,205 @@
+//
+//  FFmpegController.mm
+//  iina
+//
+//  Created by Saagar Jha on 2/16/19.
+//  Copyright © 2019 lhc. All rights reserved.
+//
+
+#import "FFmpegController.h"
+
+// Perhaps this will work someday? Until then, we need to keep around our dummy
+// interface in the header file.
+//#import "IINA-Swift.h"
+
+#import <algorithm>
+#import <memory>
+
+extern "C" {
+#import <libavcodec/avcodec.h>
+#import <libavformat/avformat.h>
+#import <libavutil/avutil.h>
+#import <libavutil/imgutils.h>
+#import <libswscale/swscale.h>
+}
+
+#define DEFAULT_THUMBNAIL_COUNT 100
+#define DEFAULT_THUMBNAIL_WIDTH 240.0
+
+#define __ACTUALLY_CONCATENATE(a, b) a##b
+#define __CONCATENATE(a, b) __ACTUALLY_CONCATENATE(a, b)
+#define DEFER(cleanup) std::shared_ptr<void> __CONCATENATE(__defer_, __LINE__)(nullptr, cleanup)
+
+@implementation FFmpegController (Bridge)
+
+- (BOOL)synchronouslyGenerateThumbnailsForFileAtPath:(NSString *)path {
+  AVFormatContext *formatContext = NULL;
+  if (avformat_open_input(&formatContext, path.fileSystemRepresentation, nullptr, nullptr) < 0) {
+    return NO;
+  }
+  DEFER([&](...) {
```

Comment:
What about putting `[&](...)` into the macro?

Reply:
I've currently chosen to leave it out to prevent putting too much "magic" into the macro (so it works like an actual "defer" that takes a full closure). But I'd be fine with hoisting the capture and parameter lists into the macro–I'd assume you're looking for something like this at call site?

```objective-c++
DEFER({
	// code
});
```
Another choice is to use Objective-C blocks, which are layout-compatible with C++ lambdas and would save two characters ;)

```
DEFER(^(...) {
	// code
});
```

Reply:
```C++
DEFER({
	// code
});
```
I prefer this one, however it's on your choice.

Reply:
Since it's only being used in this file, I'll hoist it up into the macro. If we ever need this anywhere else I think we should go for the more general solution.